### PR TITLE
extending ingress url format

### PR DESCRIPTION
**Pull request by jinxingwang**
Issue created: https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/issues/1007

### CODE REVIEW

**Comment by gongx** [comment]
@jinxingwang We might have the backward compatibility issue since the ingress URL format has been changed.
@liyinan926 Thoughts?

**Comment by jinxingwang** [comment]
There should be no backward compatibility issue, I only extended the feature.

**Comment by gongx** [comment]
@liyinan926 Please review this PR.

File: pkg/controller/sparkapplication/sparkui.go [hunk 1]
```diff
@@ -31,17 +31,19 @@ import (
 
 	"github.com/GoogleCloudPlatform/spark-on-k8s-operator/pkg/apis/sparkoperator.k8s.io/v1beta2"
 	"github.com/GoogleCloudPlatform/spark-on-k8s-operator/pkg/config"
+	"net/url"
```
Comment (liyinan926) [inline pkg/controller/sparkapplication/sparkui.go:34]: Nit: move this up to the group of imports for built-in libs.

File: pkg/controller/sparkapplication/sparkui.go [hunk 2]
```diff
@@ -61,9 +63,15 @@ type SparkIngress struct {
 }
 
 func createSparkUIIngress(app *v1beta2.SparkApplication, service SparkService, ingressURLFormat string, kubeClient clientset.Interface) (*SparkIngress, error) {
-	ingressURL := getSparkUIingressURL(ingressURLFormat, app.GetName())
+	ingressURL := getSparkUIingressURL(ingressURLFormat, app.GetName(), app.GetNamespace())
 	ingressResourceAnnotations := getIngressResourceAnnotations(app)
 	ingressTlsHosts := getIngressTlsHosts(app)
+	//adding http:// so url.Paese can function correctly
```
Comment (liyinan926) [inline pkg/controller/sparkapplication/sparkui.go:69]: What if the URL already has the `http://` prefix?. Also a typo here `Paese `.

**Comment by jinxingwang** [comment]
@liyinan926 Please take look again, thanks!

**Comment by jacobhjkim** [comment]
Hi, awesome PR. I was going to do this myself. Just a question, why does the CRD change so much?

**Comment by jinxingwang** [comment]
@jacobhjkim Thanks. I think the CRD change may due to the previous PRs that didn't have it up to date. 
If you are also trying to bind multiple spark UI under the same URL look up the settings needed here:
https://stackoverflow.com/questions/56368948/change-root-path-for-spark-web-ui

**Review by liyinan926 — APPROVED** [review]
LGTM with one question.

File: pkg/controller/sparkapplication/sparkui.go [hunk 3]
```diff
@@ -61,9 +63,22 @@ type SparkIngress struct {
...
-	ingressURL := getSparkUIingressURL(ingressURLFormat, app.GetName())
+	ingressURL := getSparkUIingressURL(ingressURLFormat, app.GetName(), app.GetNamespace())
 	ingressResourceAnnotations := getIngressResourceAnnotations(app)
 	ingressTlsHosts := getIngressTlsHosts(app)
+
+	parsedURL, err := url.Parse(ingressURL)
+	if err != nil {
+		return nil, err
+	}
+	if parsedURL.Scheme == "" {
```
[the remark this answers is not in the document]
Comment (jinxingwang) [reply pkg/controller/sparkapplication/sparkui.go:74]: This is to solve the problem where URL parser can't parse the host & path correctly when there is no schema, so I am using the parser to check if the URL has a schema, if not add a schema to parse.

**Comment by jinxingwang** [comment]
Thank you very much!

Will(Jinxing Wang)
425-974-9158 | wjx101220@hotmail.com

________________________________

**Comment by kishorv06** [comment]
@jinxingwang Great work here. I also wanted to use path based routing instead of DNS based for the spark jobs. But I can't get it working. I am getting redirected to the `/jobs` path.

> @jacobhjkim Thanks. I think the CRD change may due to the previous PRs that didn't have it up to date.
> If you are also trying to bind multiple spark UI under the same URL look up the settings needed here:
> https://stackoverflow.com/questions/56368948/change-root-path-for-spark-web-ui

I saw your comment here, but still can't figure out how to do it with `spark-on-k8s-operator`. Can you elaborate a little and possibly include it in the User Guide ?

**Comment by jinxingwang** [comment]
@kishorv06 Hi, I am happy to help you with the spark UI work. 
on my StackOverflow post. 
step 1 will be a sparkapplication spec change to add a sparkconfig `spark.ui.proxyBase: /foo`
Like this: https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/cf3446097f47047ab46a076ebdb1f71ee9428f84/docs/user-guide.md#specifying-spark-configuration

step2: will be sparkapplicaiton spec change to add annotation on the ingress when it is created
doc here: https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/bd1b7ec1f20d9a8bd991ddb335b908fab2cc3316/manifest/crds/sparkoperator.k8s.io_sparkapplications.yaml#L3653

step3: same as step 2, add the annotation to the ingress. 

Let me know if this is helping, the doc on how to get spark UI works, in my opinion, don't really belong to here, because spark operator is mostly config how spark works. it should be engineer's chose how they want to config sparkApplication to config spark to work base on their specific cases. 
That's why spark operator only docs on you can config these, but don't really care how to get it working.

**Comment by kishorv06** [comment]
Hi @jinxingwang, I followed your suggestion, but it is causing too many redirects. These are my configuration.
Application:
```yaml
  sparkConf:
    spark.ui.proxyBase: /foo
  sparkUIOptions:
    ingressAnnotations:
      nginx.ingress.kubernetes.io/proxy-redirect-from: http://$host/
      nginx.ingress.kubernetes.io/proxy-redirect-to: http://$host/foo/
      nginx.ingress.kubernetes.io/rewrite-target: /$2
```
Spark operator:
```yaml
ingressUrlFormat: "example.com/foo"
```
Result:
```bash
curl -vvv "http://example.com/foo/jobs"
*   Trying 192.168.2.102:80...
* TCP_NODELAY set
* Connected to example.com (192.168.2.102) port 80 (#0)
> GET /foo/jobs HTTP/1.1
> Host: example.com
> User-Agent: curl/7.68.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 302 Found
< Server: nginx/1.17.10
< Date: Tue, 20 Oct 2020 06:54:19 GMT
< Content-Length: 0
< Connection: keep-alive
< Cache-Control: no-cache, no-store, must-revalidate
< X-Frame-Options: SAMEORIGIN
< X-XSS-Protection: 1; mode=block
< X-Content-Type-Options: nosniff
< Location: http://example.com/foo/jobs/
< 
* Connection #0 to host example.com left intact
```

Thanks in advance.

**Comment by kishorv06** [comment]
After a lot of trial and error, finally got the path based routing to work on spark-operator. These are the configurations I have used.
Spark operator:
```yaml
ingressUrlFormat: "example.com/foo(/|$)(.*)"
```
Spark Application:
```yaml
  sparkConf:
    spatk.ui.proxyRedirectUri: http://example.com
    spark.ui.proxyBase: /foo
  sparkUIOptions:
    ingressAnnotations:
      nginx.ingress.kubernetes.io/rewrite-target: /$2
```

Note: Executor tab is empty. I think it is a known bug as mentioned here: https://github.com/jupyterhub/jupyter-server-proxy/issues/57#issuecomment-699163115

**Comment by jinxingwang** [comment]
Nice, I guess I had an assumption which you have Nginx installed. that's where ```

      nginx.ingress.kubernetes.io/proxy-redirect-from: http://$host/
      nginx.ingress.kubernetes.io/proxy-redirect-to: http://$host/foo/
```
being useful.

**Comment by andyzheung** [comment]
> @kishorv06 Hi, I am happy to help you with the spark UI work. on my StackOverflow post. step 1 will be a sparkapplication spec change to add a sparkconfig `spark.ui.proxyBase: /foo` Like this: https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/cf3446097f47047ab46a076ebdb1f71ee9428f84/docs/user-guide.md#specifying-spark-configuration
> 
> step2: will be sparkapplicaiton spec change to add annotation on the ingress when it is created doc here:
> 
> https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/bd1b7ec1f20d9a8bd991ddb335b908fab2cc3316/manifest/crds/sparkoperator.k8s.io_sparkapplications.yaml#L3653
> 
> step3: same as step 2, add the annotation to the ingress.
> 
> Let me know if this is helping, the doc on how to get spark UI works, in my opinion, don't really belong to here, because spark operator is mostly config how spark works. it should be engineer's chose how they want to config sparkApplication to config spark to work base on their specific cases. That's why spark operator only docs on you can config these, but don't really care how to get it working.

Hi， I am try to use kong as ingress-controller，and follow this guide, but seem can't work..
Is any suggestion?
1、operator:
ingressUrlFormat: "spark-ui.private-cloud.xxx.com/sparkjob(/|$)(.*)"
2、sparkapplication:
sparkConf:
    spatk.ui.proxyRedirectUri: http://spark-ui.private-cloud.xxx.com
    spark.ui.proxyBase: /sparkjob
  sparkUIOptions:
    ingressAnnotations:
      konghq.com/strip-path: "true"
      kubernetes.io/ingress.class: "kong"
      #nginx.ingress.kubernetes.io/rewrite-target: /spark-pi

but apply, get the ingress i find:
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  annotations:
    konghq.com/strip-path: "true"
    kubernetes.io/ingress.class: kong
    nginx.ingress.kubernetes.io/rewrite-target: /$2
.....
spec:
  rules:
  - host: spark-ui.private-cloud.xxx.com
    http:
      paths:
      - backend:
          service:
            name: spark-pi-ui-svc
            port:
              number: 4040
        path: /sparkjob(/|$)(.*)(/|$)(.*)
        pathType: ImplementationSpecific

**Pull request merged**